runtime/common/src/crowdsale.rs

```diff
@@ -0,0 +1,476 @@
+
+use codec::{Encode, Decode};
+use sp_runtime::{RuntimeDebug, DispatchResult, DispatchError};
+use sp_runtime::traits::{Bounded, Saturating, Zero};
+use frame_support::{decl_event, decl_storage, decl_module, decl_error, ensure};
+use frame_support::traits::{
+	EnsureOrigin, IsDeadAccount, Currency, ExistenceRequirement, VestingSchedule, Get
+};
+use sp_core::sr25519;
+
+/// Configuration trait.
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+	/// Balances Pallet
+	type Currency: Currency<Self::AccountId>;
+	/// Vesting Pallet
+	type VestingSchedule: VestingSchedule<Self::AccountId, Moment=Self::BlockNumber, Currency=Self::Currency>;
+	/// The number of blocks that the "locked" dots will be locked for.
+	type VestingTime: Get<Self::BlockNumber>;
+	/// The origin allowed to set account status.
+	type ValidityOrigin: EnsureOrigin<Self::Origin>;
+	/// The origin allowed to make final payments.
+	type PaymentOrigin: EnsureOrigin<Self::Origin>;
+}
+
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+/// The kind of a statement an account needs to make for a claim to be valid.
+#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, RuntimeDebug)]
+pub enum AccountValidity {
+	/// Account is not valid.
+	Invalid,
+	/// Account is pending validation.
+	Pending,
+	/// Account is valid with a low contribution amount.
+	ValidLow,
+	/// Account is valid with a high contribution amount.
+	ValidHigh,
+	/// Account has completed the crowdsale process.
+	Completed,
+}
+
+impl Default for AccountValidity {
+	fn default() -> Self {
+		AccountValidity::Invalid
+	}
+}
+
+impl AccountValidity {
+	fn is_valid(&self) -> bool {
+		match self {
+			Self::Invalid => false,
+			Self::Pending => false,
+			Self::ValidLow => true,
+			Self::ValidHigh => true,
+			Self::Completed => false,
+		}
+	}
+
+	fn max_amount<T: Trait>(&self) -> BalanceOf<T> {
+		match self {
+			Self::Invalid => Zero::zero(),
+			Self::Pending => Zero::zero(),
+			Self::ValidLow => 100_000.into(),
+			Self::ValidHigh => BalanceOf::<T>::max_value(),
+			Self::Completed => Zero::zero(),
+		}
+	}
+}
+
+/// All information about an account regarding the crowdsale.
+#[derive(Encode, Decode, Default, Clone, Eq, PartialEq, RuntimeDebug)]
+pub struct AccountStatus<Balance> {
+	validity: AccountValidity,
+	balance: Balance,
+	locked: bool,
+	statement_kind: StatementKind,
+	signature: Vec<u8>,
+}
+
+/// The kind of a statement an account needs to make for a claim to be valid.
+#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, RuntimeDebug)]
+pub enum StatementKind {
+	/// Statement required to be made by the regular sale participants.
+	Regular,
+	/// Statement required to be made by high value sale participants.
+	High,
+}
+
+impl StatementKind {
+	/// Convert this to the (English) statement it represents.
+	fn to_text(self) -> &'static [u8] {
```

Comment:
Maybe this could be done with associated consts on `trait Trait`?

Reply:
Yes, def can do this! I think we will only have one statement too, so this should be quite easy